File: src/ObjectManager/Object.Tes/FilePacks/Records/045-QUST.Quest.cs
using OA.Core;
using System.Collections.Generic;

namespace OA.Tes.FilePacks.Records
{
    public class QUSTRecord : Record
    {
        public struct DATAField
        {
            public byte Flags;
            public byte Priority;

            public DATAField(UnityBinaryReader r, int dataSize)
            {
                Flags = r.ReadByte();
                Priority = r.ReadByte();
            }
        }

        public override string ToString() => $"QUST: {EDID.Value}";
        public STRVField EDID { get; set; } // Editor ID
        public STRVField FULL; // Item Name
        public FILEField ICON; // Icon
        public DATAField DATA; // Icon
        public FMIDField<SCPTRecord> SCRI; // Script Name
        public SCPTRecord.SCHRField SCHR; // Script Data
        public BYTVField SCDA; // Compiled Script
        public STRVField SCTX; // Script Source
        public List<FMIDField<Record>> SCROs = new List<FMIDField<Record>>(); // Global variable reference

        public override bool CreateField(UnityBinaryReader r, GameFormatId format, string type, int dataSize)
        {
            switch (type)
            {
                case "EDID": EDID = new STRVField(r, dataSize); return true;
                case "FULL": FULL = new STRVField(r, dataSize); return true;
                case "ICON": ICON = new FILEField(r, dataSize); return true;
                case "DATA": DATA = new DATAField(r, dataSize); return true;
                case "SCRI": SCRI = new FMIDField<SCPTRecord>(r, dataSize); return true;
                case "CTDA": r.SkipBytes(dataSize); return true;
                case "INDX": r.SkipBytes(dataSize); return true;
                case "QSDT": r.SkipBytes(dataSize); return true;
                case "CNAM": r.SkipBytes(dataSize); return true;
                case "QSTA": r.SkipBytes(dataSize); return true;
                case "SCHR": SCHR = new SCPTRecord.SCHRField(r, dataSize); return true;
                case "SCDA": SCDA = new BYTVField(r, dataSize); return true;
                case "SCTX": SCTX = new STRVField(r, dataSize); return true;
                case "SCRO": SCROs.Add(new FMIDField<Record>(r, dataSize)); return true;
                default: return false;
            }
        }
    }
}

File: src.mac/ObjectManager/ObjectManager/Object.Tes/FilePacks/Records/045-QUST.Quest.swift
    //
//  QUSTRecord.swift
//  ObjectManager
//
//  Created by Sky Morey on 5/28/18.
//  Copyright © 2018 Sky Morey. All rights reserved.
//

public class QUSTRecord: Record {
    public struct DATAField {
        public let flags: UInt8
        public let priority: UInt8

        init(_ r: BinaryReader, _ dataSize: Int) {
            flags = r.readByte()
            priority = r.readByte()
        }
    }

    public override var description: String { return "QUST: \(EDID)" }
    public var EDID: STRVField = STRVField_empty // Editor ID
    public var FULL: STRVField!  // Item Name
    public var ICON: FILEField!  // Icon
    public var DATA: DATAField!  // Icon
    public var SCRI: FMIDField<SCPTRecord>! // Script Name
    public var SCHR: SCPTRecord.SCHRField! // Script Data
    public var SCDA: BYTVField! // Compiled Script
    public var SCTX: STRVField! // Script Source
    public var SCROs = [FMIDField<Record>]() // Global variable reference
    
    override func createField(_ r: BinaryReader, for format: GameFormatId, type: String, dataSize: Int) -> Bool {
        switch type {
        case "EDID": EDID = r.readSTRV(dataSize)
        case "FULL": FULL = r.readSTRV(dataSize)
        case "ICON": ICON = r.readSTRV(dataSize)
        case "DATA": DATA = DATAField(r, dataSize)
        case "SCRI": SCRI = FMIDField<SCPTRecord>(r, dataSize)
        case "CTDA": r.skipBytes(dataSize)
        case "INDX": r.skipBytes(dataSize)
        case "QSDT": r.skipBytes(dataSize)
        case "CNAM": r.skipBytes(dataSize)
        case "QSTA": r.skipBytes(dataSize)
        case "SCHR": SCHR = SCPTRecord.SCHRField(r, dataSize)
        case "SCDA": SCDA = r.readBYTV(dataSize)
        case "SCTX": SCTX = r.readSTRV(dataSize)
        case "SCRO": SCROs.append(FMIDField<Record>(r, dataSize))
        default: return false
        }
        return true
    }
}
